### PR TITLE
export_as_txt poc

### LINKED ISSUE #61: add `export_as_txt` like rtables
currently, the export functionality is very limited. To export a paginated listing need extra work of manually call "paginate" and `print` functions. something like `export_as_txt` and ~`export_as_pdf`~ like what we have in rtables are needed.

Joe: discussed with Liming. We will focus on the export_as_txt functionality, and leave txt file to pdf to downstream

### PR DESCRIPTION
**Pull request by shajoezhu**
close #61 

lpp is not yet working

functions from rtables, can be shared, consider moving to formatters

 - .do_inset
 - .paste_no_na 

# todos
- [blocked by] - same pagination utilities with rtables, need to work on pagination, see #43 
- [x] - add div line
- [see #81 ] - \s\n replace to \f @clarkliming to check
- [x] - check line break within title  
- [future work] page number, to be confirmed with @clarkliming and @barnett11 https://github.com/insightsengineering/rlistings/issues/80

see https://github.com/Roche/rtables/issues/491 and https://github.com/insightsengineering/formatters/issues/96

### CODE REVIEW

**Comment by shajoezhu** [comment]
Brilliant! Thanks @gmbecker , we will consider the page number in the next release then.

File: R/listing_export.R [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                          page_type = NULL,
+                          landscape = FALSE,
+                          pg_width = page_dim(page_type)[if (landscape) 2 else 1],
+                          pg_height = page_dim(page_type)[if (landscape) 1 else 2],
+                          font_family = "Courier",
+                          font_size = 8, # grid parameters
+                          paginate = .need_pag(page_type, pg_width, pg_height, lpp, cpp),
+                          cpp = NULL,
+                          lpp = NULL,
+                          ..., page_break = "\\f",
```
Comment (clarkliming) [inline R/listing_export.R:39]: it is `\f` instead of `\\f` by the way

Comment (shajoezhu) [reply R/listing_export.R:39]: use "\\f" will output "\f", double slash for the encoding. :) I will address this as a seperate PR per discussion we had this morning

File: tests/testthat/test-listing_dsl01.R [hunk 2]
```diff
@@ -55,13 +47,4 @@ testthat::test_that("DSL01 listing is produced correctly", {
...
   testthat::expect_identical(prov_footer(result), expected_prov_footer)
-
-  # Test DSL01 attributes
-  expected_rfnotes <- list()
-  expected_nlheader <- 3
-  expected_nrheader <- 1
-
-  testthat::expect_identical(mf_rfnotes(result_matrix), expected_rfnotes)
-  testthat::expect_equal(mf_nlheader(result_matrix), expected_nlheader)
-  testthat::expect_equal(mf_nrheader(result_matrix), expected_nrheader)
```
Comment (Melkiades) [inline tests/testthat/test-listing_dsl01.R:66]: Why not these?

File: tests/testthat/test-listing_dsl01.R [hunk 3]
```diff
@@ -18,32 +18,24 @@ testthat::test_that("DSL01 listing is produced correctly", {
     DISCONT = "Discontinued\nEarly from Study?"
   )
 
-  result <- testthat::expect_message(as_listing(
+  testthat::expect_message(as_listing(
```
Comment (Melkiades) [inline tests/testthat/test-listing_dsl01.R:21]:
```suggestion
  testthat::expect_message(result <- as_listing(
```

File: tests/testthat/test-listing_dsl01.R [hunk 4]
```diff
@@ -18,32 +18,24 @@ testthat::test_that("DSL01 listing is produced correctly", {
...
-      "No", "Yes", "No", "No", "Yes", "No", "No"
-    ),
-    nrow = 13
+  result <- as_listing(
+    dsl01,
+    disp_cols = names(dsl01),
+    main_title = "Listing of Patients with Study Drug Withdrawn Due to Adverse Events",
+    subtitles = "Population: All Patients",
+    main_footer = c("Program: xxxx", "Output: xxxx"),
+    prov_footer = "Page 1 of 1"
```
Comment (Melkiades) [inline tests/testthat/test-listing_dsl01.R:36]:
```suggestion
```

File: DESCRIPTION [hunk 5]
```diff
@@ -25,10 +25,11 @@ Suggests:
     dplyr,
     knitr,
     lifecycle,
-    testthat
+    testthat (>= 3.0)
 VignetteBuilder: 
     knitr
 Config/Needs/website: insightsengineering/nesttemplate
+Config/testthat/edition:3
```
Comment (gmbecker) [inline DESCRIPTION:32]: This is not something that should be happening as an aside as part of another PR its a (much) larger discussion that should be had separately, and certainly not this close before release

File: R/listing_export.R [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+  if (!is.null(file)) {
+    cat(res, file = file)
+  } else {
+    res
+  }
+}
+
+
+.do_inset <- function(x, inset) {
```
Comment (gmbecker) [inline R/listing_export.R:138]: This function already exists in formatters, it doesn't belong here. Also, all you should need to do is set the inset on the matrix form, during construction, there's no reason to be doing this here (it does not happen anywhere in rtables, for example)

File: R/rlistings_methods.R [hunk 7]
```diff
@@ -207,6 +207,50 @@ setMethod(
...
+  # txt_head <- .paste_no_na(names(x), collapse = gap_str)
+  #
+  if (is.null(widths)) {
+    widths <- propose_column_widths(x)
+  }
+  ncchar <- sum(widths) + (length(widths) - 1) * col_gap
+
+  div <- substr(strrep(hsep, ncchar), 1, ncchar)
+
+  inset <- 0L
```
Comment (gmbecker) [inline R/rlistings_methods.R:234]: if inset is hardcoded to 0L it isn't doing anything anyway...

**Review by gmbecker — CHANGES_REQUESTED** [review]

**Comment by shajoezhu** [comment]
closing this as new pr raised in https://github.com/insightsengineering/rlistings/pull/95

**Pull request closed without merging**